docs/source/structured_config.rst

```diff
@@ -257,13 +260,44 @@ OmegaConf verifies at runtime that your Lists contains only values of the correc
 
 Dictionaries
 ^^^^^^^^^^^^
-Dictionaries are supported as well. Keys must be strings, ints or enums, and values can
-be any of any type supported by OmegaConf (Any, int, float, bool, str and Enums as well
+Dictionaries are supported via annotation of structured config fields with ``typing.Dict``.
+Keys must be typed as one of ``str``, ``int``, ``Enum``, ``float``, or ``bool``. Values can
+be any of the types supported by OmegaConf (``Any``, ``int``, ``float``, ``bool``, ``str`` and ``Enum`` as well
 as arbitrary Structured configs)
 
+.. doctest::
+
+    >>> from typing import Dict
+    >>> @dataclass
+    ... class DictExample:
+    ...     # Typed dict keys are strings; values can be typed as Any, int, float, bool, str and Enums or
+    ...     # arbitrary Structured configs
+    ...     ints: Dict[str, int] = field(default_factory=lambda: {"a": 10, "b": 20, "c": 30})
+    ...     bools: Dict[str, bool] = field(default_factory=lambda: {"Uno": True, "Zoro": False})
+    ...     users: Dict[str, User] = field(default_factory=lambda: {"omry": User(name="omry")})
+
+Like with Lists, the types of values contained in Dicts are verified at runtime.
+
+.. doctest::
+
+    >>> conf: DictExample = OmegaConf.structured(DictExample)
+
+    >>> # Okay, 10 is an int
+    >>> conf.ints["d"] = 10
+    >>> # Okay, "20" can be converted to an int
+    >>> conf.ints["e"] = "20"
```

Comment:
We are thinking about deprecating the automatic conversion on assignment because it triggers a typing error (#459).
Don't document the automatic conversion, just the straight forward valid assignment and the hard error.

Reply:
Done in https://github.com/omry/omegaconf/pull/706/commits/a23fbac1f0674aaf6b0c1811e642b8175902b847.